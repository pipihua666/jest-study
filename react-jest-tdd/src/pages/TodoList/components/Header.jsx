import React, { Component } from 'react'

class Header extends Component {
  state = {
    value: ''
  }

  handleChange = e => {
    this.setState({
      value: e.target.value
    })
  }

  handleKeyDown = e => {
    if (e.keyCode === 13 && this.state.value) {
      this.props.addUndoItem(this.state.value)
      this.setState({ value: '' })
    }
  }

  render() {
    return (
      <div>
        <input
          onKeyUp={this.handleKeyDown}
          data-test="input"
          value={this.state.value}
          onChange={this.handleChange}
        />
      </div>
    )
  }
}

export default Header
