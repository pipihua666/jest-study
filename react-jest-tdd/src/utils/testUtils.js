export const findElementByTag = (wrapper, tag) => {
  return wrapper.find(`[data-test="${tag}"]`)
}

export const findElementByComponentName = (wrapper, componentName) => {
  return wrapper.find(componentName)
}
