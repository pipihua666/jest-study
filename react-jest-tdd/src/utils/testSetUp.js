import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import Header from '../pages/TodoList/components/Header'

Enzyme.configure({ adapter: new Adapter() })
