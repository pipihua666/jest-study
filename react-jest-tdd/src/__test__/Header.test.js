import { shallow } from 'enzyme'

import Header from '../pages/TodoList/components/Header'

import { findElementByTag } from '../utils/testUtils'

it('快照', () => {
  const wrapper = shallow(<Header />)
  expect(wrapper.html()).toMatchSnapshot()
})

it('Header有一个input框', () => {
  const wrapper = shallow(<Header />)
  const inputEl = findElementByTag(wrapper, 'input')
  expect(inputEl.length).toBe(1)
})

it('input框的初始值为空', () => {
  const wrapper = shallow(<Header />)
  const inputEl = findElementByTag(wrapper, 'input')
  expect(inputEl.prop('value')).toBe('')
})

it('input是一个受控组件', () => {
  const wrapper = shallow(<Header />)
  const inputEl = findElementByTag(wrapper, 'input')
  const inputValue = 'Hello Pipihua!'
  inputEl.simulate('change', {
    target: {
      value: inputValue
    }
  })
  expect(wrapper.state('value')).toBe(inputValue)
  // 为什么要获取input元素呢？
  // 因为在onChange之后旧的input value是没有改变的
  const newInputEl = findElementByTag(wrapper, 'input')
  expect(newInputEl.prop('value')).toBe(inputValue)
})

it('input框输入回车时，如果input无内容，无操作', () => {
  const fn = jest.fn()
  const wrapper = shallow(<Header addUndoItem={fn} />)
  const inputEl = findElementByTag(wrapper, 'input')
  wrapper.setState({ value: '' })
  inputEl.simulate('keyUp', {
    keyCode: 13
  })
  expect(fn).not.toHaveBeenCalled()
})

it('input框输入回车时，如果input有内容，执行函数操作', () => {
  const fn = jest.fn()
  const wrapper = shallow(<Header addUndoItem={fn} />)
  const inputEl = wrapper.find("[data-test='input']")
  const inputValue = '皮皮花学习jest!'
  wrapper.setState({ value: inputValue })
  inputEl.simulate('keyUp', {
    keyCode: 13
  })
  expect(fn).toHaveBeenCalled()
  expect(fn).toHaveBeenCalledWith(inputValue)
})

it('input框输入回车时，如果input有内容，清空输入框的值', () => {
  const fn = jest.fn()
  const wrapper = shallow(<Header addUndoItem={fn} />)
  const inputEl = wrapper.find("[data-test='input']")
  const inputValue = '皮皮花学习jest!'
  wrapper.setState({ value: inputValue })
  inputEl.simulate('keyUp', {
    keyCode: 13
  })
  expect(wrapper.state('value')).toBe('')
})
