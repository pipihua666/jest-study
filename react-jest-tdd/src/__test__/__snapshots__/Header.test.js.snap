// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照 1`] = `"<div><input data-test=\\"input\\" value=\\"\\"/></div>"`;
